### PR TITLE
ARROW-16168: [C++][CMake] Use target to add include paths

### PR DESCRIPTION
**Pull request by kou**
We can remove "include_directories(SYSTEM)" by this.

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly

**Review by cyb70289 — APPROVED** [review]

**Comment by kou** [comment]
@github-actions crossbow submit amazon* *ubuntu-18.04* *ubuntu-20.04*

**Comment by kou** [comment]
@github-actions crossbow submit *ubuntu-18.04*

**Comment by kou** [comment]
@github-actions crossbow submit verify-rc-source-cpp-linux-ubuntu-18.04-amd64

**Comment by kou** [comment]
@github-actions crossbow submit verify-rc-source-cpp-linux-ubuntu-*-amd64

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Comment by pitrou** [comment]
@kou Should this be updated for git master? Is there anything left to do otherwise?

**Comment by kou** [comment]
> Should this be updated for git master?

Yes.

> Is there anything left to do otherwise?

I need to confirm that the current failures are fixed by rebasing on the master. If there are unknown failures, I need to fix them in this pull request. (I prioritized 8.0.0 release related tasks over this.)

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Comment by kou** [comment]
@github-actions crossbow submit almalinux-8-amd64

**Comment by kou** [comment]
@github-actions crossbow submit almalinux-8-amd64 ubuntu-bionic-amd64 test-ubuntu-20.04-cpp-thread-sanitizer wheel-windows-cp310-amd64

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Comment by kou** [comment]
This is ready to merge.
I'll merge this tomorrow if there is no objection.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -692,10 +692,15 @@ endif()
 #
 
 # Libraries to link statically with libarrow.so
-set(ARROW_LINK_LIBS)
-set(ARROW_STATIC_LINK_LIBS)
+set(ARROW_LINK_LIBS arrow::flatbuffers arrow::hadoop)
+set(ARROW_STATIC_LINK_LIBS arrow::flatbuffers arrow::hadoop)
 set(ARROW_STATIC_INSTALL_INTERFACE_LIBS)
 
+if(TARGET Boost::headers)
```
Comment (pitrou) [inline cpp/CMakeLists.txt:699]: This looks a bit weird, shouldn't the condition be based on enabled Arrow components instead?

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -808,6 +813,16 @@ if(ARROW_WITH_RE2)
...
 endif()
 
+if(ARROW_WITH_RAPIDJSON)
+  list(APPEND ARROW_LINK_LIBS rapidjson::rapidjson)
+  list(APPEND ARROW_STATIC_LINK_LIBS rapidjson::rapidjson)
+endif()
+
+if(TARGET xsimd)
+  list(APPEND ARROW_LINK_LIBS xsimd)
+  list(APPEND ARROW_STATIC_LINK_LIBS xsimd)
```
Comment (pitrou) [inline cpp/CMakeLists.txt:823]: AFAIK, xsimd is header-only, is this necessary?

File: cpp/cmake_modules/BuildUtils.cmake [hunk 3]
```diff
@@ -59,78 +43,25 @@ function(ADD_THIRDPARTY_LIB LIB_NAME)
...
-    endif()
-    message(STATUS "Added static library dependency ${AUG_LIB_NAME}: ${ARG_STATIC_LIB}")
-    if(ARG_INCLUDE_DIRECTORIES)
-      set_target_properties(${AUG_LIB_NAME} PROPERTIES INTERFACE_INCLUDE_DIRECTORIES
-                                                       "${ARG_INCLUDE_DIRECTORIES}")
-    endif()
-  elseif(ARG_SHARED_LIB)
-    set(AUG_LIB_NAME "${LIB_NAME}_shared")
-    add_library(${AUG_LIB_NAME} SHARED IMPORTED)
-
```
Comment (pitrou) [inline cpp/cmake_modules/BuildUtils.cmake:114]: Nice simplification here!

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -966,31 +1002,72 @@ else()
...
 
 if(ARROW_BOOST_REQUIRED)
+  if(ARROW_BOOST_USE_SHARED)
+    # Find shared Boost libraries.
+    set(Boost_USE_STATIC_LIBS OFF)
+    set(BUILD_SHARED_LIBS_KEEP ${BUILD_SHARED_LIBS})
+    set(BUILD_SHARED_LIBS ON)
+  else()
+    # Find static boost headers and libs
+    # TODO Differentiate here between release and debug builds
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1012]: Can you open a JIRA for this TODO?

**Comment by pitrou** [comment]
@kou Is the AppVeyor failure unrelated?

**Comment by kou** [comment]
> @kou Is the AppVeyor failure unrelated?

Yes. It should be fixed by https://github.com/apache/arrow/pull/13101 .

File: cpp/CMakeLists.txt [hunk 5]
```diff
@@ -692,10 +692,15 @@ endif()
 #
 
 # Libraries to link statically with libarrow.so
-set(ARROW_LINK_LIBS)
-set(ARROW_STATIC_LINK_LIBS)
+set(ARROW_LINK_LIBS arrow::flatbuffers arrow::hadoop)
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/CMakeLists.txt:695]: I see. Can you add that last sentence as a comment?

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (pitrou) [reply cpp/CMakeLists.txt:823]: Ah, ok. Thanks.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 4 above)
Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1012]: Yes, let's just remove it.

File: cpp/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (kou) [reply cpp/CMakeLists.txt:699]: Done.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 4 above)
Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Done.

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Comment by kou** [comment]
+1

**Pull request closed without merging**